File: packages/contracts/src/chart/birth-input.ts
import { z } from 'zod';

const calendarKinds = ['gregorian', 'lunar'] as const;
const birthInputSources = ['user-entered', 'imported', 'test-fixture'] as const;
const chartSexValues = ['male', 'female', 'unknown'] as const;

export const birthDateSchema = z.object({
  year: z.number().int().min(1).max(9999),
  month: z.number().int().min(1).max(12),
  day: z.number().int().min(1).max(31),
  isLeapMonth: z.boolean().nullable(),
});

export const birthTimeSchema = z
  .object({
    hour: z.number().int().min(0).max(23).nullable(),
    minute: z.number().int().min(0).max(59).nullable(),
    isUnknown: z.boolean(),
  })
  .superRefine((value, ctx) => {
    if (value.isUnknown) {
      return;
    }

    if (value.hour === null || value.minute === null) {
      ctx.addIssue({
        code: 'custom',
        message: 'Known birth time requires both hour and minute.',
      });
    }
  });

export const manualCoordinatesSchema = z.object({
  latitude: z.number().min(-90).max(90),
  longitude: z.number().min(-180).max(180),
  timezone: z.string().min(1),
});

export const birthPlaceSchema = z
  .object({
    label: z.string().trim().min(1).nullable(),
    manual: manualCoordinatesSchema.nullable(),
  })
  .superRefine((value, ctx) => {
    if (value.label || value.manual) {
      return;
    }

    ctx.addIssue({
      code: 'custom',
      message: 'Birth place requires a label or manual coordinates.',
    });
  });

export const birthInputSchema = z
  .object({
    calendar: z.enum(calendarKinds),
    date: birthDateSchema,
    time: birthTimeSchema,
    sexOrGenderForChart: z.enum(chartSexValues),
    place: birthPlaceSchema,
    locale: z.string().min(2),
    source: z.enum(birthInputSources),
  })
  .superRefine((value, ctx) => {
    if (value.calendar === 'lunar' && value.date.isLeapMonth === null) {
      ctx.addIssue({
        code: 'custom',
        path: ['date', 'isLeapMonth'],
        message: 'Lunar birth input must say whether the month is leap.',
      });
    }

    if (value.calendar === 'gregorian' && value.date.isLeapMonth !== null) {
      ctx.addIssue({
        code: 'custom',
        path: ['date', 'isLeapMonth'],
        message: 'Gregorian birth input cannot set leap-month metadata.',
      });
    }
  });

export type BirthDate = z.infer<typeof birthDateSchema>;
export type BirthTime = z.infer<typeof birthTimeSchema>;
export type ManualCoordinates = z.infer<typeof manualCoordinatesSchema>;
export type BirthPlace = z.infer<typeof birthPlaceSchema>;
export type BirthInput = z.infer<typeof birthInputSchema>;

File: apps/mobile/lib/features/charts/data/models/birth_input.dart
import 'package:json_annotation/json_annotation.dart';

part 'birth_input.g.dart';

@JsonSerializable(explicitToJson: true)
class BirthDate {
  final int year;
  final int month;
  final int day;
  final bool? isLeapMonth;

  BirthDate({
    required this.year,
    required this.month,
    required this.day,
    this.isLeapMonth,
  });

  factory BirthDate.fromJson(Map<String, dynamic> json) => _$BirthDateFromJson(json);
  Map<String, dynamic> toJson() => _$BirthDateToJson(this);
}

@JsonSerializable(explicitToJson: true)
class BirthTime {
  final int? hour;
  final int? minute;
  final bool isUnknown;

  BirthTime({
    this.hour,
    this.minute,
    required this.isUnknown,
  });

  factory BirthTime.fromJson(Map<String, dynamic> json) => _$BirthTimeFromJson(json);
  Map<String, dynamic> toJson() => _$BirthTimeToJson(this);
}

@JsonSerializable(explicitToJson: true)
class ManualCoordinates {
  final double latitude;
  final double longitude;
  final String timezone;

  ManualCoordinates({
    required this.latitude,
    required this.longitude,
    required this.timezone,
  });

  factory ManualCoordinates.fromJson(Map<String, dynamic> json) => _$ManualCoordinatesFromJson(json);
  Map<String, dynamic> toJson() => _$ManualCoordinatesToJson(this);
}

@JsonSerializable(explicitToJson: true)
class BirthPlace {
  final String? label;
  final ManualCoordinates? manual;

  BirthPlace({
    this.label,
    this.manual,
  });

  factory BirthPlace.fromJson(Map<String, dynamic> json) => _$BirthPlaceFromJson(json);
  Map<String, dynamic> toJson() => _$BirthPlaceToJson(this);
}

@JsonSerializable(explicitToJson: true)
class BirthInput {
  final String calendar; // 'gregorian' or 'lunar'
  final BirthDate date;
  final BirthTime time;
  final String sexOrGenderForChart; // 'male', 'female', 'unknown'
  final BirthPlace place;
  final String locale;
  final String source; // 'user-entered', etc.

  BirthInput({
    required this.calendar,
    required this.date,
    required this.time,
    required this.sexOrGenderForChart,
    required this.place,
    required this.locale,
    required this.source,
  });

  factory BirthInput.fromJson(Map<String, dynamic> json) => _$BirthInputFromJson(json);
  Map<String, dynamic> toJson() => _$BirthInputToJson(this);
}
